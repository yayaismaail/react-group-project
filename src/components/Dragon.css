.container {
  display: flex;
}

.dragon-img {
  max-width: 20vw;
  margin: 20px;
}

.reserve-btn {
  background-color: blue;
  color: white;
  border: none;
  border-radius: 4px;
  height: 30px;
  width: 150px;
  padding: 10px 5px;
  display: flex;
  justify-content: center;
  align-items: center;
}

.cancel-btn {
  background-color: whitesmoke;
  color: grey;
  border: 1px solid grey;
  border-radius: 4px;
  height: 30px;
  width: 150px;
  padding: 10px 5px;
  display: flex;
  justify-content: center;
  align-items: center;
}

.badge {
  background-color: rgb(19, 181, 187);
  color: white;
  border-radius: 4px;
  height: 20px;
}
