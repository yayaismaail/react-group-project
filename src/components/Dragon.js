import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { reserveDragon, cancelReservation } from '../redux/dragons/dragonsSlice';
import './Dragon.css';

const Dragon = ({ dragons }) => {
  const dispatch = useDispatch();

  const handleReserve = (id) => {
    dispatch(reserveDragon(id));
    localStorage.setItem(`dragon-${id}`, 'reserved');
  };

  const handleCancelReservation = (id) => {
    dispatch(cancelReservation(id));
    localStorage.removeItem(`dragon-${id}`);
  };

  const isReserved = (id) => localStorage.getItem(`dragon-${id}`) === 'reserved';

  return (
    <div>
      {dragons.map((dragon) => (
        <div key={dragon.id} className="container">
          {dragon.flickr_images && dragon.flickr_images.length > 0 && (
            <img
              src={dragon.flickr_images[0]}
              alt={dragon.name}
              className="dragon-img"
            />
          )}
          <div className="text-wrapper">
            <div>
              <h2>{dragon.name}</h2>
              <p>
                {isReserved(dragon.id) && <span className="badge">Reserved</span>}
                {' '}
                {dragon.description}
              </p>
            </div>
            {isReserved(dragon.id) ? (
              <button
                type="button"
                className="cancel-btn"
                onClick={() => handleCancelReservation(dragon.id)}
              >
                Cancel reservation
              </button>
            ) : (
              <button
                type="button"
                className="reserve-btn"
                onClick={() => handleReserve(dragon.id)}
              >
                Reserve Dragon
              </button>
            )}
          </div>
        </div>
      ))}
    </div>
  );
};

export default Dragon;

Dragon.propTypes = {
  dragons: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      description: PropTypes.string.isRequired,
      flickr_images: PropTypes.arrayOf(PropTypes.string),
    }),
  ).isRequired,
};
