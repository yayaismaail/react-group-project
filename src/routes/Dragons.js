import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { fetchAllDragons } from '../redux/dragons/dragonsSlice';
import Dragon from '../components/Dragon';

const Dragons = () => {
  const dispatch = useDispatch();
  const dragons = useSelector((state) => state.dragons.dragons);
  const isLoading = useSelector((state) => state.dragons.isLoading);
  const ifSucceed = useSelector((state) => state.dragons.ifSucceed);

  useEffect(() => {
    dispatch(fetchAllDragons());
  }, [dispatch]);

  if (isLoading) {
    return <div>Loading...</div>;
  }

  return (
    <div>
      {ifSucceed ? <Dragon dragons={dragons} /> : null}
    </div>
  );
};

export default Dragons;
