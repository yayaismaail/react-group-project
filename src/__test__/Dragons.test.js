import React from 'react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import Dragons from '../routes/Dragons';
import Dragon from '../components/Dragon';

describe('all dragons components render', () => {
  it('dragons component renders properly', () => {
    const userObj = jest.mock('../routes/Dragons', () => {
      renderer.create(<Dragons />).toJSON();
    });
    expect(userObj).toMatchSnapshot();
  });

  it('dragon component renders properly', () => {
    const userObj = jest.mock('../components/Dragon', () => {
      renderer.create(<Dragon />).toJSON();
    });
    expect(userObj).toMatchSnapshot();
  });
});
