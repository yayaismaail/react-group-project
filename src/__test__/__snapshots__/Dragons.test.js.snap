// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all dragons components render dragon component renders properly 1`] = `
Object {
  "advanceTimersByTime": [Function],
  "advanceTimersToNextTimer": [Function],
  "autoMockOff": [Function],
  "autoMockOn": [Function],
  "clearAllMocks": [Function],
  "clearAllTimers": [Function],
  "createMockFromModule": [Function],
  "deepUnmock": [Function],
  "disableAutomock": [Function],
  "doMock": [Function],
  "dontMock": [Function],
  "enableAutomock": [Function],
  "fn": [Function],
  "genMockFromModule": [Function],
  "getRealSystemTime": [Function],
  "getTimerCount": [Function],
  "isMockFunction": [Function],
  "isolateModules": [Function],
  "mock": [Function],
  "mocked": [Function],
  "requireActual": [Function],
  "requireMock": [Function],
  "resetAllMocks": [Function],
  "resetModules": [Function],
  "restoreAllMocks": [Function],
  "retryTimes": [Function],
  "runAllImmediates": [Function],
  "runAllTicks": [Function],
  "runAllTimers": [Function],
  "runOnlyPendingTimers": [Function],
  "setMock": [Function],
  "setSystemTime": [Function],
  "setTimeout": [Function],
  "spyOn": [Function],
  "unmock": [Function],
  "unstable_mockModule": [Function],
  "useFakeTimers": [Function],
  "useRealTimers": [Function],
}
`;

exports[`all dragons components render dragons component renders properly 1`] = `
Object {
  "advanceTimersByTime": [Function],
  "advanceTimersToNextTimer": [Function],
  "autoMockOff": [Function],
  "autoMockOn": [Function],
  "clearAllMocks": [Function],
  "clearAllTimers": [Function],
  "createMockFromModule": [Function],
  "deepUnmock": [Function],
  "disableAutomock": [Function],
  "doMock": [Function],
  "dontMock": [Function],
  "enableAutomock": [Function],
  "fn": [Function],
  "genMockFromModule": [Function],
  "getRealSystemTime": [Function],
  "getTimerCount": [Function],
  "isMockFunction": [Function],
  "isolateModules": [Function],
  "mock": [Function],
  "mocked": [Function],
  "requireActual": [Function],
  "requireMock": [Function],
  "resetAllMocks": [Function],
  "resetModules": [Function],
  "restoreAllMocks": [Function],
  "retryTimes": [Function],
  "runAllImmediates": [Function],
  "runAllTicks": [Function],
  "runAllTimers": [Function],
  "runOnlyPendingTimers": [Function],
  "setMock": [Function],
  "setSystemTime": [Function],
  "setTimeout": [Function],
  "spyOn": [Function],
  "unmock": [Function],
  "unstable_mockModule": [Function],
  "useFakeTimers": [Function],
  "useRealTimers": [Function],
}
`;
